### PR TITLE
[eclipse/xtext#1249] use bom for version management

### PR DESCRIPTION
**Pull request by szarnekow**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -22,10 +23,17 @@ subprojects {
 	version = rootProject.version
 	
 	repositories {
+		mavenLocal()
```
Comment (cdietrich) [inline build.gradle]: am not sure if i like that or if we should pass a `maven{url "somepath"}` here

File: org.eclipse.xtend.lib.gwt.test/.project [hunk 2]
```diff
@@ -16,7 +16,7 @@
 			</arguments>
 		</buildCommand>
 		<buildCommand>
-			<name>com.google.gwt.eclipse.core.gwtProjectValidator</name>
+			<name>com.gwtplugins.gwt.eclipse.core.gwtProjectValidator</name>
```
Comment (cdietrich) [inline org.eclipse.xtend.lib.gwt.test/.project]: is this change intended?

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+		<junit5-version>5.1.0</junit5-version>
+		<log4j-version>1.2.17</log4j-version>
+		<!-- // When adjusting the `lsp4j` version here, do not forget to do the 
+			followings: // - Update the version in `IdeProjectDescriptor#pom`. // - Update 
+			the test expectations by running `CliWizardIntegrationTest#main`. // - Update 
+			the versions in the `/org.eclipse.xtext.ide/META-INF/MANIFEST.MF` and `/org.eclipse.xtext.ide.tests/META-INF/MANIFEST.MF` 
+			files. -->
+		<lsp4j-version>0.6.0</lsp4j-version>
+		<xpand-version>2.0.0</xpand-version>
+		<guava-version>[14.0,21.0]</guava-version>
```
Comment (cdietrich) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: this change is intentional, correct?

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+			</dependency>
+			<dependency>
+				<groupId>com.google.guava</groupId>
+				<artifactId>guava-gwt</artifactId>
+				<version>${guava-version}</version>
+			</dependency>
+			<dependency>
+				<groupId>org.antlr</groupId>
+				<artifactId>antlr-runtime</artifactId>
+				<version>3.2</version>
```
Comment (cdietrich) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: no variable for that?

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+			</dependency>
+			<dependency>
+				<groupId>org.eclipse.emf</groupId>
+				<artifactId>org.eclipse.emf.codegen.ecore</artifactId>
+				<version>${emfEcore-version}</version>
+			</dependency>
+			<dependency>
+				<groupId>org.eclipse.platform</groupId>
+				<artifactId>org.eclipse.core.commands</artifactId>
+				<version>3.9.0</version>
```
Comment (cdietrich) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: no variable for that?

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+			</dependency>
+			<dependency>
+				<groupId>org.eclipse.platform</groupId>
+				<artifactId>org.eclipse.equinox.common</artifactId>
+				<version>${equinoxCommon-version}</version>
+			</dependency>
+			<dependency>
+				<groupId>org.eclipse.platform</groupId>
+				<artifactId>org.eclipse.equinox.preferences</artifactId>
+				<version>3.7.0</version>
```
Comment (cdietrich) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: no variable here?

File: org.eclipse.xtend.lib.gwt.test/.project (hunk 2 above)
Comment (kthoms) [reply org.eclipse.xtend.lib.gwt.test/.project]: this happened automatically

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 3 above)
Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: you mean managing guava version?

Comment (cdietrich) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: i mean the new range

Comment (cdietrich) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: for guava

Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: I would rather like to have a fixed version, but didn't you mention somewhere that it is intentional to have a range?

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 4 above)
Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: This version won't change ever, so why having a var? I'd leave it as is.

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 3 above)
Comment (cdietrich) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: see https://github.com/eclipse/xtext-lib/blob/3d3e03de9fec9c148576aaa2879999d13371144b/gradle/versions.gradle#L11

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 5 above)
Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: Actually I don't like to have too much variables that are only used once.

Comment (szarnekow) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: Without the variables, it's more verbose to override a version from the bom in a client pom, isn't it?

Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: Correct, that's a point.
However, I find it difficult to relate the variables to the actually bundle they belong to. How about using a property name scheme like `<artifactId>-version`, e.g. `org.eclipse.core.commands-version`?

File: org.eclipse.xtend.lib.gwt.test/.project (hunk 2 above)
Comment (kthoms) [reply org.eclipse.xtend.lib.gwt.test/.project]: I'll revert this one.

File: build.gradle (hunk 1 above)
Comment (kthoms) [reply build.gradle]: too much magic when using a URL here. It is always a bit harder to know to what it actually resolves. The local maven repo is clear to that. The Jenkins script makes sure it us a private one.

File: 2-gradle-build.sh [hunk 7]
```diff
@@ -0,0 +1,8 @@
+./gradlew \
+  clean cleanGenerateXtext build createLocalMavenRepo \
+  -Dmaven.repo.local=$(pwd)/.m2/repository \
+  -PcompileXtend=true \
+  -PJENKINS_URL=https://ci.eclipse.org/xtext/ \
```
Comment (szarnekow) [inline 2-gradle-build.sh]: hardcoded jenkins url?

File: 1-install-bom.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+mvn \
+  -f releng/org.eclipse.xtext.dev-bom \
+  --batch-mode \
+  --update-snapshots \
+  -Dmaven.repo.local=.m2/repository \
+  -DJENKINS_URL=https://ci.eclipse.org/xtext/ \
```
Comment (szarnekow) [inline 1-install-bom.sh]: Hardcoded jenkins url? How does this work?

File: Jenkinsfile [hunk 9]
```diff
@@ -1,24 +1,49 @@
...
+		stage('Gradle Build') {
+			steps {
+				sh 'sh ./2-gradle-build.sh'
+				step([$class: 'JUnitResultArchiver', testResults: '**/build/test-results/test/*.xml'])
+			}
+		}
+		
+		stage('Maven Build') {
+			steps {
+				dir('.m2/repository/org/eclipse/xtext') { deleteDir() }
```
Comment (szarnekow) [inline Jenkinsfile]: Won't this delete the xtext-bom?

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+		<log4j-version>1.2.17</log4j-version>
+		<!-- // When adjusting the `lsp4j` version here, do not forget to do the 
+			followings: // - Update the version in `IdeProjectDescriptor#pom`. // - Update 
+			the test expectations by running `CliWizardIntegrationTest#main`. // - Update 
+			the versions in the `/org.eclipse.xtext.ide/META-INF/MANIFEST.MF` and `/org.eclipse.xtext.ide.tests/META-INF/MANIFEST.MF` 
+			files. -->
+			
+		<!-- Eclipse Platform -->
+		<!-- Variable naming scheme: -->
+		<!-- * skip 'org.eclipse' prefix, use hyphen separator instead of . -->
```
Comment (szarnekow) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: Not sure I like this. If we'd use org.eclipse.foobar-version, searching for the bundle name will yield the version as a grep-result, too.

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+			<!-- Eclipse Platform -->
+			<dependency>
+				<groupId>org.eclipse.emf</groupId>
+				<artifactId>org.eclipse.emf.codegen</artifactId>
+				<version>${emf-codegen-version}</version>
+			</dependency>
+			<dependency>
+				<groupId>org.eclipse.emf</groupId>
+				<artifactId>org.eclipse.emf.codegen.ecore</artifactId>
+				<version>${emf-ecore-version}</version>
```
Comment (szarnekow) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: Is this correct?

File: releng/pom.xml [hunk 12]
```diff
@@ -10,6 +10,7 @@
 	<properties>
 		<tycho-version>1.3.0</tycho-version>
 		<root-dir>${basedir}/..</root-dir>
+		<maven.deploy.skip>true</maven.deploy.skip>
```
Comment (szarnekow) [inline releng/pom.xml]: TODO revert

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+		<plugins>
+			<plugin>
+				<artifactId>maven-resources-plugin</artifactId>
+				<version>3.1.0</version>
+				<executions>
+					<execution>
+						<id>copy-resource-one</id>
+						<phase>install</phase>
+						<goals>
+							<goal>copy-resources</goal>
```
Comment (szarnekow) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: Why is this here?

**Comment by kthoms** [comment]
Good points. Will update when having time.

**Comment by szarnekow** [comment]
> Good points. Will update when having time.

I pushed some changes already. Please be careful with friendly pushes :)

File: Jenkinsfile (hunk 9 above)
Comment (kthoms) [reply Jenkinsfile]: right, copy/paste issue

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 10 above)
Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: agreed, that would be better. will change.

Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: took that because it was already started that way

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 13 above)
Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: This copies the BOM during the `install` phase to the target location that is archived . This avoids the misuse of the `deploy` goal and would allow to use the deploy goal for a real deployment to a remote repo.

File: releng/pom.xml (hunk 12 above)
Comment (kthoms) [reply releng/pom.xml]: OK

**Comment by kthoms** [comment]
I'm not happy with the BOM's artifactId `xtext-dev.bom`. What should users expect from the `dev` part? It is _the_ Xtext BOM. IMHO it should be rather
```
<artifactId>xtext-bom</artifactId>
```

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+	<properties>
+		<!-- ======== -->
+		<!-- Versions -->
+		<!-- ======== -->
+		<!-- 3rd party -->
+		<antlr.runtime-version>3.2</antlr.runtime-version>
+		<args4j-version>2.33</args4j-version>
+		<asm-version>7.0</asm-version>
+		<error_prone_annotations-version>2.0.15</error_prone_annotations-version>
+		<guava-version>[14.0,22.0]</guava-version>
```
Comment (szarnekow) [inline releng/org.eclipse.xtext.dev-bom/pom.xml]: Is this correct? Should it be [14.0,22.0) ?

**Comment by szarnekow** [comment]
> I'm not happy with the BOM's artifactId `xtext-dev.bom`. What should users expect from the `dev` part? It is _the_ Xtext BOM. IMHO it should be rather
> 
> ```
> <artifactId>xtext-bom</artifactId>
> ```
> And maybe also the folder be better `org.eclipse.xtext.bom` then?

The dev-bom was mainly for the development of xtext itself, so for xtext-dev. For users, we should eventually provide an xtext-bom. But that is a different story.

File: releng/org.eclipse.xtext.dev-bom/pom.xml [hunk 15]
```diff
@@ -98,7 +98,7 @@
 			<dependency>
 				<groupId>org.antlr</groupId>
 				<artifactId>antlr-runtime</artifactId>
-				<version>${antlr-runtime-version}</version>
+				<version>${antlr.runtime-version}</version>
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: Not critical though. We can change that later.

**Comment by kthoms** [comment]
Offline agreed with @szarnekow that the artifactId should be `xtext-dev-bom` to avoid mixup of seperators. This will affect changing downstream PRs. Will update now.

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: will be updated

File: org.eclipse.xtend.lib.gwt.test/build.gradle [hunk 16]
```diff
@@ -32,3 +32,7 @@ gwt {
 test {
 	dependsOn compileGwt
 }
+
+tasks.withType(Javadoc) {
```
Comment (szarnekow) [inline org.eclipse.xtend.lib.gwt.test/build.gradle]: Why is that?

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 14 above)
Comment (szarnekow) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: @cdietrich

**Comment by kthoms** [comment]
I think we can further manage some scopes and exclusions in the BOM, too. E.g. use `test` scope for test libraries and `runtime` scope for codegen dependencies (e.g. MWE runtime, EMF codegen) and exclusions for log4j.

File: org.eclipse.xtend.lib.gwt.test/build.gradle (hunk 16 above)
Comment (szarnekow) [reply org.eclipse.xtend.lib.gwt.test/build.gradle]: Should we keep it like this?

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 14 above)
Comment (szarnekow) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: 21.0 instead?

File: org.eclipse.xtend.lib.gwt.test/build.gradle (hunk 16 above)
Comment (kthoms) [reply org.eclipse.xtend.lib.gwt.test/build.gradle]: if you don't mind...?

File: releng/org.eclipse.xtext.dev-bom/pom.xml (hunk 14 above)
Comment (kthoms) [reply releng/org.eclipse.xtext.dev-bom/pom.xml]: I think you are right. 22.0 should be excluded and usually we use the excluded version as upper bound. So `[14.0,22.0)` should be better here.

File: org.eclipse.xtend.lib.gwt.test/build.gradle (hunk 16 above)
Comment (szarnekow) [reply org.eclipse.xtend.lib.gwt.test/build.gradle]: Ok
